src/mesh/mesh_refinement_ops.hpp

```diff
@@ -0,0 +1,243 @@
+  return 0.5 * (SIGN(gxm) + SIGN(gxp)) * std::min(std::abs(gxm), std::abs(gxp));
+}
+} // namespace util
+
+template <int DIM>
+struct RestrictCellAverage {
+  KOKKOS_FORCEINLINE_FUNCTION static void
+  Do(const int l, const int m, const int n, const int ck, const int cj, const int ci,
+     const IndexRange &ckb, const IndexRange &cjb, const IndexRange &cib,
+     const IndexRange &kb, const IndexRange &jb, const IndexRange &ib,
+     const Coordinates_t &coords, const Coordinates_t &coarse_coords,
+     const ParArray6D<Real> &coarse, const ParArray6D<Real> &fine) {
+    const int i = (ci - cib.s) * 2 + ib.s;
+    int j = jb.s;
+    if (DIM > 1) {
+      j = (cj - cjb.s) * 2 + jb.s;
+    }
+    int k = kb.s;
+    if (DIM > 2) {
+      k = (ck - ckb.s) * 2 + kb.s;
+    }
+    // JMM: If dimensionality is wrong, accesses are out of bounds. Only
+    // access cells if dimensionality is correct.
+    const Real vol000 = coords.Volume(k, j, i);
+    const Real vol001 = coords.Volume(k, j, i + 1);
+    const Real fine000 = fine(l, m, n, k, j, i);
+    const Real fine001 = fine(l, m, n, k, j, i + 1);
+    Real vol010, vol011, vol100, vol101, vol110, vol111;
+    Real fine010, fine011, fine100, fine101, fine110, fine111;
+    vol010 = vol011 = vol100 = vol101 = vol110 = vol111 = 0;
+    fine010 = fine011 = fine100 = fine101 = fine110 = fine111 = 0;
+    if (DIM > 1) { // TODO(c++17) make constexpr
+      vol010 = coords.Volume(k, j + 1, i);
+      vol011 = coords.Volume(k, j + 1, i + 1);
+      fine010 = fine(l, m, n, k, j + 1, i);
+      fine011 = fine(l, m, n, k, j + 1, i + 1);
+    }
+    if (DIM > 2) { // TODO(c++17) make constexpr
+      vol100 = coords.Volume(k + 1, j, i);
+      vol101 = coords.Volume(k + 1, j, i + 1);
+      vol110 = coords.Volume(k + 1, j + 1, i);
+      vol111 = coords.Volume(k + 1, j + 1, i + 1);
+      fine100 = fine(l, m, n, k + 1, j, i);
+      fine101 = fine(l, m, n, k + 1, j, i + 1);
+      fine110 = fine(l, m, n, k + 1, j + 1, i);
+      fine111 = fine(l, m, n, k + 1, j + 1, i + 1);
+    }
+    // KGF: add the off-centered quantities first to preserve FP
+    // symmetry
+    const Real tvol =
+        ((vol000 + vol010) + (vol001 + vol011)) + ((vol100 + vol110) + (vol101 + vol111));
+    coarse(l, m, n, ck, cj, ci) =
+        (((fine000 * vol000 + fine010 * vol010) + (fine001 * vol001 + fine011 * vol011)) +
+         ((fine100 * vol100 + fine110 * vol110) +
+          (fine101 * vol101 + fine111 * vol111))) /
+        tvol;
+  }
+};
+
+template <int DIM>
+struct ProlongateCellMinMod {
+  KOKKOS_FORCEINLINE_FUNCTION static void
+  Do(const int l, const int m, const int n, const int k, const int j, const int i,
+     const IndexRange &ckb, const IndexRange &cjb, const IndexRange &cib,
+     const IndexRange &kb, const IndexRange &jb, const IndexRange &ib,
+     const Coordinates_t &coords, const Coordinates_t &coarse_coords,
+     const ParArray6D<Real> &coarse, const ParArray6D<Real> &fine) {
```

Comment:
what about just removing `const`?

Reply:
I think it is more of how it looks on the calling end. For a function `f(double a_in, double b_in, double& c_out)` it is unclear at first glance at the call site what might be being updated
```c++
double a, b, c; 
func(a, b, c); 
```
Switching to `f(double a_in, double b_in, double* c_out)` and calling with 
```c++
double a, b, c; 
func(a, b, &c); 
```
makes it instantly clear that `c` may be changed by `func`.

Reply:
yeah ok I can see that. I don't think it really matters here, but I don't feel strongly about it, so I'll change it.

Reply:
One thing to just put a pin in: when capturing `Kokkos::View`s, it's actually pretty important the capture be by value. But this is already all device code, so I think it won't matter here.

Reply:
Actually---I can't change the `ParArray6D`s to pointers, because the `[=]` capture in `KOKKOS_LAMBDA` automatically makes everything thats captured into a `const`.

Reply:
Well---maybe const pointers. I'll try that and see if the compiler accepts it.